Objects/typeobject.c

```diff
@@ -6971,6 +6976,12 @@ PyType_Ready(PyTypeObject *type)
         type->tp_flags |= Py_TPFLAGS_IMMUTABLETYPE;
     }
 
+    /* All immutable types must have a static valid version tag */
+    if (type->tp_flags & Py_TPFLAGS_IMMUTABLETYPE) {
+        type->tp_version_tag = next_version_tag++;
+        type->tp_flags |= Py_TPFLAGS_VALID_VERSION_TAG;
+    }
```

Comment:
tl;dr I'm on board with this.

As I mentioned in [gh-101696](https://github.com/python/cpython/issues/101696#issuecomment-1424535319), this is good as it covers non-builtin static types (AKA non-`_Py_TPFLAGS_STATIC_BUILTIN`), e.g. in community extension modules. [^1]  (The switch to `Py_TPFLAGS_IMMUTABLETYPE` seems fine.)

FWIW, the already merged fix may fit a little better with the vague idea (in my head) that we should be special-casing static types in the runtime as little as possible (or otherwise in more explicit ways, like the dedicated `_PyStaticType_Dealloc()`).  In that spirit, rather than modifying `PyType_Ready()`, I was going to suggest we could find a way to apply `_PyStaticType_Dealloc()` to all static types, not just the builtin ones, which might be a good idea regardless.  However, that feels like overkill for fixing this issue.  Plus, we're already special-casing static types here in `PyType_Ready()`.  (Perhaps my mental model is misaligned...)

Bottom line: dropping the original fix and adding the fix you have here is worth it since it handles non-builtin static types too.


[^1]: One *could* argue that non-builtin (non-core) static types should be discouraged (basically, they already are) and we shouldn't bother adding extra maintenance burden--however small--to accommodate them.  However, that's not where the community is at nor will be for a long time and there's no reason to be so hard-line about it, especially in a case like this where the additional maintenance burden is essentially zero.

Reply:
Note that this PR does _not_ special-case static types; the condition is for _immutable_ types.